corehq/pillows/synclog.py

```diff
@@ -0,0 +1,139 @@
+from corehq.apps.receiverwrapper.util import get_version_and_app_from_build_id
+from corehq.apps.users.models import LastSync, CouchUser, CommCareUser, WebUser
+from corehq.pillows.utils import update_latest_builds, filter_by_app
+from corehq.util.doc_processor.interface import BaseDocProcessor, DocumentProcessorController
+from corehq.util.doc_processor.couch import CouchDocumentProvider
+
+from dimagi.utils.parsing import string_to_utc_datetime
+
+from pillowtop.pillow.interface import ConstructedPillow
+from pillowtop.processors.interface import PillowProcessor
+from pillowtop.feed.couch import CouchChangeFeed
+from pillowtop.feed.interface import Change
+from pillowtop.checkpoints.manager import PillowCheckpoint, PillowCheckpointEventHandler
+from pillowtop.reindexer.reindexer import Reindexer
+
+from casexml.apps.phone.models import SyncLog
+from casexml.apps.phone.dbaccessors.sync_logs_by_user import get_synclogs_for_user
+
+
+def get_synclog_pillow(pillow_id='SynclogPillow'):
+    """
+    This gets a pillow which iterates through all synclogs
+    """
+    couch_db = SyncLog.get_db()
+    change_feed = CouchChangeFeed(couch_db, include_docs=True)
+    checkpoint = PillowCheckpoint('synclog', couch_db)
+    form_processor = SynclogProcessor()
+    return ConstructedPillow(
+        name=pillow_id,
+        checkpoint=checkpoint,
+        change_feed=change_feed,
+        processor=form_processor,
+        change_processed_event_handler=PillowCheckpointEventHandler(
+            checkpoint=checkpoint, checkpoint_frequency=100
+        ),
+    )
+
+
+class SynclogProcessor(PillowProcessor):
+
+    def process_change(self, pillow_instance, change):
+
+        synclog = change.get_document()
+        if not synclog:
+            return
+
+        version = None
+        app_id = None
+        try:
+            last_sync_date = string_to_utc_datetime(synclog.get('date'))
+        except ValueError:
+            return
+        build_id = synclog.get('build_id')
+        if build_id:
+            version, app_id = get_version_and_app_from_build_id(synclog.get('domain'), build_id)
+        user_id = synclog.get('user_id')
+
+        if user_id:
+            user = CouchUser.get_by_user_id(user_id)
+
+            last_sync = filter_by_app(user.reporting_metadata.last_syncs, app_id)
+
+            if last_sync is None or last_sync_date >= last_sync.sync_date:
+                if last_sync is None:
+                    last_sync = LastSync()
+                    user.reporting_metadata.last_syncs.append(last_sync)
+                last_sync.sync_date = last_sync_date
+                last_sync.build_version = version
+                last_sync.app_id = app_id
+
+                if user.reporting_metadata.last_sync_for_user is None \
+                        or last_sync_date > user.reporting_metadata.last_sync_for_user.sync_date:
+                    user.reporting_metadata.last_sync_for_user = last_sync
```

Comment:
Could extract this code into `update_last_sync` function similar to `update_latest_builds`